### DISCUSSION TITLE
Raft consensus while file uploading

### DISCUSSION THREAD

**Discussion by tahirali-csc** [seaweedfs/seaweedfs#4198] [General]
I was looking at how the raft is used especially when replicating the file to multiple servers. In particular, `seaweedfs/weed/topology/store_replicate.go` and it looks like `for` loop over all replicas instead of relying on raft to perform replication.

```go
if len(remoteLocations) > 0 { //send to other replica locations
		start := time.Now()
		err = DistributedOperation(remoteLocations, func(location operation.Location) error {
			u := url.URL{
				Scheme: "http",
				Host:   location.Url,
				Path:   r.URL.Path,
			}
```

I have seen hashicorp raft library in the code but not sure why in the replication we are doing manually for loop instead of raft. Any help or suggestion on this?

**Comment by kmlebedev** [comment]
Raft need for clustering master nodes.

**Reply by kmlebedev** [reply]
It is also not entirely clear how the raft will help? If we need to physically put a chunk in each replica? information about this, which is already requested by grpc from the leader of the master server.
